### PR TITLE
fix_: limit the maximum number of message hashes by query hash

### PR DESCRIPTION
**Pull request by richard-ramos**
Team, do you think this could be added to the release or is it too risky at this point? 

The problem of not adding this code is that if the missing message verification is enabled there's no limit for the number of missing message hashes to request and that can increase the load of the storenodes while with this change we reduce the number of message hashes per query to 50.

The equivalent PR for `develop` branch will be created later once https://github.com/waku-org/go-waku/pull/1190 is merged

### CODE REVIEW

**Review by chaitanyaprem — APPROVED** [review]
LGTM

**Comment by chaitanyaprem** [comment]
> Team, do you think this could be added to the release or is it too risky at this point?

I think we can add to release if we dogfood the changes and see no issues. Especially if it helps improve store node performance.

**Comment by richard-ramos** [comment]
Sounds good, I'll create a dogfooding PR and hopefully we can do a quick dogfooding session early next week to get this merged ASAP! :rocket:

**Review by kaichaosun — APPROVED** [review]
LGTM

**Review by ilmotta — APPROVED** [review]
> Team, do you think this could be added to the release or is it too risky at this point?

@richard-ramos, code looks good, but before merging directly into the release branch we can ask the mobile QA team @status-im/mobile-qa to have a look (we just need to quickly create a mobile PR pointing to this branch).

It would be less risky if `FetchHistory` was covered by tests.

**Comment by richard-ramos** [comment]
I have created the following test PRs:
- https://github.com/status-im/status-desktop/pull/16079
- https://github.com/status-im/status-mobile/pull/21021

File: wakuv2/missing_messages.go [hunk 1]
```diff
@@ -236,44 +237,63 @@ func (w *Waku) fetchMessagesBatch(missingHistoryRequest TopicInterest, batchFrom
...
-	}, logger, "retrieving missing messages")
-	if err != nil {
-		if !errors.Is(err, context.Canceled) {
-			logger.Error("storenode not available", zap.Error(err))
+	wg := sync.WaitGroup{}
+	// Split into batches
+	for i := 0; i < len(missingHashes); i += maxHashQueryLength {
+		j := i + maxHashQueryLength
+		if j > len(missingHashes) {
+			j = len(missingHashes)
```
Comment (igor-sirotin) [inline wakuv2/missing_messages.go]:
```suggestion
		j = math.Min(j, len(missingHashes))
```

**Review by igor-sirotin — COMMENTED** [review]
But isn't pagination already setting this limit? 🤔 
And why is it better to do a few parallel requests (to the same store node), rather then sequential?

**Comment by richard-ramos** [comment]
> But isn't pagination already setting this limit? 🤔

The query we do kinda looks like this:
```
SELECT * 
FROM messages
WHERE messageHash IN (msgHash1, msgHash2, .... msgHashN)
LIMIT 100
```
The `pageLimit` will limit the results __after__ the filtering per `messageHash` is done. I.E. if you pass 300 message hashes as an `IN` condition, the database will attempt to find all those messages first before doing the limit, and will have to do the same for each page of the results if we pass a cursor.  Since the `messageHash` is unique, it's more efficient to just find and return a smaller resultset, since it will not have to find all the records first to later just return a subset of them (It also means less load for the DB).

---

Regarding sequential vs parallel, sending smaller requests concurrently I believe it is the better approach, as it can significantly reduce the overall time needed to retrieve all the data. I'm 100% not sure if nwaku will really benefit from this change but some of the benefits with this approach are:
- Reduced total latency: By sending requests in parallel, you can reduce the overall time to receive all responses. This is particularly important if each request has a significant latency (e.g., due to network delays or query processing time).
- Better utilization of resources: nwaku uses chronos, so in theory it should handle multiple concurrent operations efficiently. We also have a connection pool in postgresql. By sending requests in parallel, we should be able to take advantage of the available resources (e.g., CPU, network bandwidth, available connections) to process multiple requests simultaneously.
- Improved User Experience: users will experience faster response times as data is retrieved and processed more quickly.

**Comment by igor-sirotin** [comment]
> The pageLimit will limit the results after the filtering per messageHash is done.

But... shouldn't we change the query on server (nwaku) side then, rather than modifying clients not to overload servers? 🤔 
I mean, just limit the `mshHash` list before putting it into query.

**Comment by igor-sirotin** [comment]
> some of the benefits with this approach are: ...

Ok, I guess it makes sense in this case 👍

Though, again, some of this points sounds like we're trying to make clients to use servers more efficient, while we could make servers smarter and work efficient even with dumb clients 😄  (e.g., to utilize postgresql connection pool, the SQL request could be split into multiple requests on server side).

**Comment by richard-ramos** [comment]
Closing due to deciding to not go with this PR for 2.30 as described in https://github.com/status-im/status-mobile/pull/21021#issuecomment-2286168579

**Pull request closed without merging**